Bug: Negated patterns are considered explicit match

### Environment

Node version: v22.14.0
npm version: v10.9.2
Local ESLint version: v9.28.0
Global ESLint version: no
Operating System: windows


### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```js
export default [
    {
        files: ["!foo.js"],
        rules: {
            "no-undef": 2
        }
    }
];
```
</details>

```js
// foo.js
foo
```


### What did you expect to happen?

I'd expect only .js, .mjs, and .cjs files to be linted, and the config object to apply to all files except `foo.js`.

### What actually happened?

The `"!foo.js"` pattern is considered an explicit match, so it enables linting of all files.

```
$ npx eslint .

C:\projects\tmp\tmp\package-lock.json
  2:9  error  Parsing error: Unexpected token :

C:\projects\tmp\tmp\package.json
  2:9  error  Parsing error: Unexpected token :

✖ 2 problems (2 errors, 0 warnings)
```

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/stackblitz-starters-skkgpzj7

### Participation

- [x] I am willing to submit a pull request for this issue.

### Additional comments

_No response_

--- a/package.json
+++ b/package.json
@@ -106,7 +106,7 @@
   "dependencies": {
     "@eslint-community/eslint-utils": "^4.2.0",
     "@eslint-community/regexpp": "^4.12.1",
-    "@eslint/config-array": "^0.20.0",
+    "@eslint/config-array": "^0.20.1",
     "@eslint/config-helpers": "^0.2.1",
     "@eslint/core": "^0.14.0",
     "@eslint/eslintrc": "^3.3.1",

--- a/tests/lib/eslint/eslint.js
+++ b/tests/lib/eslint/eslint.js
@@ -2925,6 +2925,107 @@ describe("ESLint", () => {
 				});
 			});
 
+			describe("Globbing based on configs with negated patterns and arrays in `files`", () => {
+				// https://github.com/eslint/eslint/issues/19813
+				it("should not include custom extensions when negated pattern is specified in `files`", async () => {
+					eslint = new ESLint({
+						flags,
+						cwd: getFixturePath("file-extensions"),
+						overrideConfigFile: true,
+						overrideConfig: [
+							{
+								files: ["!foo.js"],
+							},
+							{
+								files: ["!foo.jsx"],
+							},
+							{
+								files: ["!foo.ts"],
+							},
+							{
+								files: ["!g.tsx"],
+							},
+						],
+					});
+					const results = await eslint.lintFiles(["."]);
+
+					// should not include d.jsx, f.ts, and other extensions that are not linted by default
+					assert.strictEqual(results.length, 4);
+					assert.deepStrictEqual(
+						results.map(({ filePath }) => path.basename(filePath)),
+						["a.js", "b.mjs", "c.cjs", "eslint.config.js"],
+					);
+				});
+
+				it("should not include custom extensions when negated pattern is specified in an array in `files`", async () => {
+					eslint = new ESLint({
+						flags,
+						cwd: getFixturePath("file-extensions"),
+						overrideConfigFile: true,
+						overrideConfig: [
+							{
+								files: [["*", "!foo.js"]],
+							},
+							{
+								files: [["!foo.js", "*"]],
+							},
+							{
+								files: [["*", "!foo.ts"]],
+							},
+							{
+								files: [["!foo.ts", "*"]],
+							},
+							{
+								files: [["*", "!g.tsx"]],
+							},
+							{
+								files: [["!g.tsx", "*"]],
+							},
+						],
+					});
+					const results = await eslint.lintFiles(["."]);
+
+					// should not include d.jsx, f.ts, and other extensions that are not linted by default
+					assert.strictEqual(results.length, 4);
+					assert.deepStrictEqual(
+						results.map(({ filePath }) => path.basename(filePath)),
+						["a.js", "b.mjs", "c.cjs", "eslint.config.js"],
+					);
+				});
+
+				// https://github.com/eslint/eslint/issues/19814
+				it("should include custom extensions when matched by a non-universal pattern specified in an array in `files`", async () => {
+					eslint = new ESLint({
+						flags,
+						cwd: getFixturePath("file-extensions", ".."),
+						overrideConfigFile: true,
+						overrideConfig: [
+							{
+								files: [["**/*.jsx", "file-extensions/*"]],
+							},
+							{
+								files: [["file-extensions/*", "**/*.ts"]],
+							},
+						],
+					});
+					const results = await eslint.lintFiles(["file-extensions"]);
+
+					// should include d.jsx and f.ts, but not other extensions that are not linted by default
+					assert.strictEqual(results.length, 6);
+					assert.deepStrictEqual(
+						results.map(({ filePath }) => path.basename(filePath)),
+						[
+							"a.js",
+							"b.mjs",
+							"c.cjs",
+							"d.jsx",
+							"eslint.config.js",
+							"f.ts",
+						],
+					);
+				});
+			});
+
 			it("should report zero messages when given a '**' pattern with a .js and a .js2 file", async () => {
 				eslint = new ESLint({
 					flags,

